src/framework/core/js/ModelTransformationTransforms.js

```diff
@@ -591,14 +589,65 @@ var fluid = fluid || fluid_2_0;
     };
 
     fluid.defaults("fluid.transforms.indexOf", {
-        gradeNames: "fluid.standardTransformFunction"
+        gradeNames: ["fluid.standardTransformFunction", "fluid.lens"],
+        invertConfiguration: "fluid.transforms.indexOf.invert"
     });
 
     fluid.transforms.indexOf = function (value, transformSpec) {
+        var offset;
+        if (transformSpec.offset !== undefined) {
+            offset = Number(transformSpec.offset);
+            if (isNaN(offset) || offset !== parseInt(offset, 10)) {
+                fluid.fail("indexOf requires the value of \"offset\" to be an integer or a string that can be converted to an integer. " + transformSpec.offset + " is invalid.");
+            }
+        }
         var array = fluid.makeArray(transformSpec.array);
-        return array.indexOf(value);
+        var originalIndex = array.indexOf(value);
+        return originalIndex === -1 && transformSpec.notFound ? transformSpec.notFound : (typeof (offset) === "number" ? originalIndex + offset : originalIndex);
     };
 
+    fluid.transforms.indexOf.invert = function (transformSpec, transformer) {
+        var togo = fluid.transforms.getCommonInvertedSpec(transformSpec, transformer);
+        togo.type = "fluid.transforms.dereference";
+        return togo;
+    };
+
+    fluid.defaults("fluid.transforms.dereference", {
+        gradeNames: ["fluid.standardTransformFunction", "fluid.lens"],
+        invertConfiguration: "fluid.transforms.dereference.invert"
+    });
+
+    fluid.transforms.dereference = function (value, transformSpec) {
+        if (typeof (value) !== "number") {
+            fluid.fail("dereference requires \"value\" to be a number. " + value + " is invalid.");
+        }
+        if (transformSpec.offset !== undefined) {
+            var offset = Number(transformSpec.offset);
+            if (isNaN(offset) || offset !== parseInt(offset, 10)) {
+                fluid.fail("dereference requires the value of \"offset\" to be an integer or a string that can be converted to an integer. " + transformSpec.offset + " is invalid.");
+            }
+        }
+        var array = fluid.makeArray(transformSpec.array);
+        var index = typeof (offset) === "number" ? value + offset : value;
+        return index === -1 && transformSpec.notFound ? transformSpec.notFound : array[index];
+    };
+
+    fluid.transforms.dereference.invert = function (transformSpec, transformer) {
+        var togo = fluid.transforms.getCommonInvertedSpec(transformSpec, transformer);
+        togo.type = "fluid.transforms.indexOf";
+        return togo;
+    };
+
+    fluid.transforms.getCommonInvertedSpec = function (transformSpec, transformer) {
```

Comment:
This name should be better qualified - it needs to reflect the fact that it is somehow related to the array rules. I suggest something like fluid.transforms.invertArrayIndexation or some such thing